Content.Server/GameObjects/EntitySystems/ReactionSystem.cs

```diff
@@ -0,0 +1,113 @@
+using Robust.Shared.IoC;
+using Robust.Shared.Prototypes;
+
+namespace Content.Server.GameObjects.EntitySystems
+{
+    /// <summary>
+    /// Checks <see cref="SolutionComponent"/>s for conditions for chemical reactions and performs them if they're met.
+    /// </summary>
+    class ReactionSystem : EntitySystem
+    {
+#pragma warning disable 649
+        [Dependency] private readonly IPrototypeManager _prototypeManager;
+#pragma warning restore 649
+
+        private IReadOnlyList<ReactionPrototype> _reactions;
+        private AudioSystem _audioSystem;
+
+        public override void Initialize()
+        {
+            EntityQuery = new TypeEntityQuery(typeof(SolutionComponent));
+
+            _reactions = _prototypeManager.EnumeratePrototypes<ReactionPrototype>().ToList();
+            _audioSystem = EntitySystemManager.GetEntitySystem<AudioSystem>();
+        }
+
+        public override void Update(float frameTime)
+        {
+            //Iterate through entities with SolutionComponents
+            foreach (var entity in RelevantEntities)
+            {
+                var solution = entity.GetComponent<SolutionComponent>();
+                if (solution.CurrentVolume == 0)
+                    continue; //Skip empty solutions
+
+                //For each solution check every reaction
+                foreach (var reaction in _reactions)
+                {
+                    if (SolutionValidReaction(solution, reaction, out int unitReactions))
+                    {
+                        PerformReaction(solution, reaction, unitReactions);
+                        break; //Only perform one reaction per solution per update.
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Checks if a solution has the reactants required to cause a specified reaction.
+        /// </summary>
+        /// <param name="solution">The solution to check for reaction conditions.</param>
+        /// <param name="reaction">The reaction whose reactants will be checked for in the solution.</param>
+        /// <param name="unitReactions">The number of times the reaction can occur with the given solution.</param>
+        /// <returns></returns>
+        private bool SolutionValidReaction(SolutionComponent solution, ReactionPrototype reaction, out int unitReactions)
+        {
+            unitReactions = 10000000; //Set to some impossibly large number initially
+            foreach (var reactant in reaction.Reactants)
+            {
+                if (!solution.ContainsReagent(reactant.Key, out int reagentQuantity))
+                {
+                    return false;
+                }
+                int currentUnitReactions = reagentQuantity / reactant.Value.Amount;
+                if (currentUnitReactions < unitReactions)
+                {
+                    unitReactions = currentUnitReactions;
+                }
+            }
+            return true;
```

Comment:
As far as I can tell, this will still return true if `unitReactions` is zero. This doesn't seem very great to me.

Reply:
Yup, you're right. Turns out that can result in a bug if you have a reaction where a unit reaction is some number > 1u. So say it takes 10u sodium for the reaction, and you only have 1u sodium in the solution, it'd still return true and react. I'll fix this when I move `ReactionSystem` into solutions.